### DISCUSSION TITLE
Custom helpers in `zkevm-proverjs`: Usage, continued support and docs

### DISCUSSION THREAD

**Discussion by mooori** [0xPolygonHermez/pilcom#65] [Q&A] [answered]
For the instrumentation of zkASM that was generated from WASM we would like to rely on custom helpers in the following way.

## Helper definition

```js
class MyHelper {
    // Initiates internal state.
    constructor(config) {
        this.config = config;
        this.counter = 0;
    }

    setup() {
        // zkevm-proverjs throws an error if the helper Object doesn't have this function.
    }

    eval_increaseMyHelperCounter() {
        // Modifies internal state.
        this.counter += 1;
    }

    eval_writeMyHelperData() {
        // Writes internal state to a file.
    }
}
```

## Helper usage

```js
// Passing it to zkevm-proverjs as part of config:
let config = {
	// ...
	helpers: [
		new MyHelper(someConfig),
	]
}
```

```
;; Calling helper functions in zkASM:
$${increaseMyHelperCounter()} ;; sets this.counter to 1
$${increaseMyHelperCounter()} ;; sets this.counter to 2
$${writeMyHelperData()}       ;; writes helper state to a file
```

## Questions

1. Above works on the branch of `zkevm-proverjs` that `near/wasmtime` is using (i.e. [/feature/64bits](https://github.com/0xPolygonHermez/zkevm-proverjs/tree/feature/64bits)). Can we rely on helpers like this being supported in the future? Would it work on the latest release too?
2. It seems like there is no documentation on custom helpers, or did we just miss it? If there is none, we could write a short tutorial on setting up and using helpers. Would that be a helpful contribution?

**Comment by krlosMata** [comment, marked as answer]
## Comments
- the `setup()` function load some standard functions into the helper class, functions commonly used in helpers class
  - `setup()` standard function --> https://github.com/0xPolygonHermez/zkevm-proverjs/blob/develop/src/sm/sm_main/helpers/helpers.js#L6
  - function that are loaded dynamically on Class helpers --> https://github.com/0xPolygonHermez/zkevm-proverjs/blob/develop/src/sm/sm_main/sm_main_exec.js#L188  

## 1
- The idea is to continue following this approach and improve it
  - we now import some helpers, abstracted in a different Class, directly in the `main_executor`: https://github.com/0xPolygonHermez/zkevm-proverjs/blob/develop/src/sm/sm_main/helpers/helpers.js
  - Somehow, we are thinking in a way that each `xxx.zkasm` will have its helpers associated. Then, when you execute `xxx.zkasm` all its helpers are loaded dynamically
  - we are still figuring out how to do that in a nice and cleaner way

## 2
- yep, there is no documentation on that topic
- sure it will be a valuable contribution !!

**Reply by mooori** [reply]
## `setup` function

From the perspective of a `zkevm-proverjs` user, [failing](https://github.com/0xPolygonHermez/zkevm-proverjs/blob/c3cab39e345298407d44291cb8527e792d2eef63/src/sm/sm_main/sm_main_exec.js#L185-L187) if the helper doesn’t have a `setup` function seems a bit harsh. For example the helper I’m currently working on requires none of the standard functions passed to `setup`. Still, the helper needs to have an empty `setup() {}` function to avoid the failure.

## Custom helper docs

> sure it will be a valuable contribution !!

Great! What would be a good place for that doc? Perhaps a file like `docs/custom_helper.md` inside the `zkevm-proverjs` repo?
